common/lib/dependabot/update_checkers/version_filters.rb

```diff
@@ -9,11 +9,20 @@ module VersionFilters
       extend T::Sig
 
       sig do
-        params(
-          versions_array: T::Array[T.any(Gem::Version, T::Hash[Symbol, Gem::Version])],
-          security_advisories: T::Array[SecurityAdvisory]
-        )
-          .returns(T::Array[T.any(Gem::Version, T::Hash[Symbol, Gem::Version])])
+        # Tricky generics explanation:
+        # There's a type T that is either a Gem::Version or a Hash with a :version key
+        # The method returns an array of T
+        # So whichever is provided as input, the output will be an array of the same type.
+        type_parameters(:T)
+          .params(
+            versions_array: T::Array[
+              T.any(
+                T.all(T.type_parameter(:T), Gem::Version),
```

Comment:
can these types be named? Tagging @bdragon and @JamieMagee to weigh in

Reply:
Generic type parameters across many languages use T, U, V in that order for type names.

Commonly the Java docs are referenced as the convention:
- https://docs.oracle.com/javase/tutorial/java/generics/types.html

Here's an example in the Go source code:
- https://github.com/golang/go/blob/a3f69c778fa6c082d38f9cdfacab27714f08a343/src/cmd/cgo/internal/test/typeparam.go#L10

Example in the Cargo book using `<T>` and the other cases:
- https://doc.rust-lang.org/book/ch10-00-generics.html

C++ using <T>:
- https://en.cppreference.com/w/cpp/language/templates

Reply:
Sorry I wasn't clear - I understand the T, V, U naming convention for generics (I've used it too).

I meant to ask about `type aliases` like [this](https://github.com/dependabot/dependabot-core/pull/9232#discussion_r1516598410) example.

Reply:
Ah ok, I can make type aliases for `T::Array[Gem::Version]` and `T::Array[T.untyped]` if you think that would be helpful.

Reply:
Only if it helps with readability :) Your call.

Reply:
Worth noting, I think `T::Hash[Symbol, Gem::Version]` is actually incorrect, it's a Hash that for the key `:version` it's a `Gem::Version`, but all the other keys are not `Gem::Version` so might make sense to introduce a proper class instead which would eliminate the need for a type alias.

Reply:
Captured this info here so we can come back to it after this PR merges: https://github.com/dependabot/dependabot-core/issues/9283